types/src/ledger_info.rs

```diff
@@ -373,6 +375,187 @@ impl LedgerInfoWithPartialSignatures {
+    }
+
+    // Aggregates all the signatures, verifies the aggregate signature, and returns the aggregate signature.
+    pub fn aggregate_and_verify(
+        &mut self,
+        epoch_state: Arc<EpochState>,
+    ) -> Result<LedgerInfoWithSignatures, VerifyError> {
+        self.check_voting_power(&epoch_state.verifier)?;
+
+        let mut all_signatures = self.verified_signatures.clone();
+        for (author, signature) in self.unverified_signatures.signatures() {
+            all_signatures.add_signature(*author, signature.clone());
+        }
+
+        let aggregated_sig = epoch_state.verifier.aggregate_signatures(&all_signatures)?;
+
+        let (verified_aggregate_signature, malicious_authors) = match epoch_state
+            .verifier
+            .clone()
+            .verify_multi_signatures(self.ledger_info(), &aggregated_sig)
+        {
+            Ok(_) => {
+                for (account_address, signature) in self.unverified_signatures.signatures() {
+                    self.verified_signatures
+                        .add_signature(*account_address, signature.clone());
+                }
+                self.unverified_signatures = PartialSignatures::empty();
+                (aggregated_sig, vec![])
+            },
+            Err(_) => {
+                // Question: Should we assign min tasks per thread here for into_par_iter()?
+                let verified = self
+                    .unverified_signatures
+                    .signatures()
+                    .into_par_iter()
+                    .flat_map(|(account_address, signature)| {
+                        if epoch_state
+                            .verifier
+                            .verify(*account_address, self.ledger_info(), signature)
+                            .is_ok()
+                        {
+                            return Some((*account_address, signature.clone()));
+                        }
+                        None
+                    })
+                    .collect::<Vec<_>>();
+                for (account_address, signature) in verified {
+                    self.verified_signatures
+                        .add_signature(account_address, signature.clone());
+                    self.unverified_signatures.remove_signature(account_address);
+                }
+                let malicious_authors = self
+                    .unverified_signatures
+                    .signatures()
+                    .keys()
+                    .cloned()
+                    .collect();
+                self.unverified_signatures = PartialSignatures::empty();
+
+                let aggregated_sig = epoch_state
+                    .verifier
+                    .aggregate_signatures(&self.verified_signatures)?;
+                // epoch_state
+                //     .read()
+                //     .verifier
+                //     .verify_multi_signatures(self.ledger_info(), &aggregated_sig)?;
```

Comment:
Should still verify the multisig?

Reply:
I don't have a solid opinion here. Is it necessary to verify the aggregate of all verified signatures again?

Reply:
NVM it is unnecessary.